### ISSUE TITLE
`emit()` is nondeterministic

### ISSUE THREAD

**Issue by dcebotarenco** [instancio/instancio#905]
**Describe the bug**
emit() - emits values differently on nested lists and depth

**Expected result**
emits the same values for each node and random seed

**Actual result**

Different seeds have different results

**Additional information**
It seems like as soon as I provide enough values for the Queue of the generator in my case I have 2x2 values = 4 values, the test passes

```java
@Data
static class RootNode {
    List<Depth1Node> depth1Nodes;

    @Data
    static class Depth1Node {
        List<Depth2Node> depth2Nodes;
        Boolean isDeleted;

        @Data
        static class Depth2Node {
            Boolean isDeleted;
        }
    }
}

@Test
@Seed(3095387440446971337L) // test fails
//    @Seed(3486230662535349159L) //test passes
void emit2ValuesForEachParentNode() {
    RootNode rootNode = Instancio.of(RootNode.class)
            .generate(field(RootNode::getDepth1Nodes), gen -> gen.collection().size(2))
            .generate(field(RootNode.Depth1Node::getIsDeleted), gen -> gen.emit().items(true, false))
            .generate(field(RootNode.Depth1Node::getDepth2Nodes), gen -> gen.collection().size(2))
            .generate(field(RootNode.Depth1Node.Depth2Node::getIsDeleted), gen -> gen.emit().items(true, false))
            .create();

    assertThat(rootNode.depth1Nodes).extracting(RootNode.Depth1Node::getIsDeleted).containsExactly(true, false);
    assertThat(rootNode.depth1Nodes.stream().flatMap(d -> d.getDepth2Nodes().stream()))
            .extracting(RootNode.Depth1Node.Depth2Node::getIsDeleted).containsExactly(true, false, true, false);

}
```

- Java version: 21
- Instancio version: `4.2.0`

![image](https://github.com/instancio/instancio/assets/1030373/e56545c2-eb9a-4b6e-a02e-007901ecf96b)
![image](https://github.com/instancio/instancio/assets/1030373/3bf819b9-e984-425b-93b6-24a3867960c4)

**Comment by armandino** [comment]
@dcebotarenco thanks for the sample! I was able to reproduce it. The behaviour is a bit confusing, but it's actually working as expected. I think it's not well documented.

As you mentioned, `Depth2Node::getIsDeleted` will have 4 values, but the `emit()` generator only provides 2 values:

```java
.generate(field(RootNode.Depth1Node.Depth2Node::getIsDeleted), gen -> gen.emit().items(true, false))
```

After emitting the provided `(true, false)`, the generator will produce random values, because the default behaviour is [`whenEmptyEmitRandom()`](https://javadoc.io/doc/org.instancio/instancio-core/latest/org/instancio/generator/specs/EmitGeneratorSpec.html#whenEmptyEmitRandom()). The seed that passes the test happens to generate another `(true, false)`. Changing the generator to return `(true, false, true, false)` should fix the failing test.

I think it might a good enhancement to supporting recycling the values once they've been exhausted. I added support for this in #906. Instead of repeating `emit(...)` values, you should be able to invoke `whenEmptyRecycle()`:

```java
.generate(field(RootNode.Depth1Node.Depth2Node::getIsDeleted), gen -> gen.emit().items(true, false).whenEmptyRecycle())
```

**Comment by dcebotarenco** [comment]
Thank you, appreciated!

**Comment by armandino** [comment]
@dcebotarenco this was released in `4.3.0`, please give it a try.
